pkg/querier/block_streaming.go

```diff
@@ -0,0 +1,280 @@
+	return bqss.currSeries
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) Err() error {
+	return nil
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) Warnings() storage.Warnings {
+	return nil
+}
+
+// newBlockStreamingQuerierSeries makes a new blockQuerierSeries. Input labels must be already sorted by name.
+func newBlockStreamingQuerierSeries(lbls labels.Labels, seriesIdxStart, seriesIdxEnd int, streamReader chunkStreamReader) *blockStreamingQuerierSeries {
+	return &blockStreamingQuerierSeries{
+		labels:         lbls,
+		seriesIdxStart: seriesIdxStart,
+		seriesIdxEnd:   seriesIdxEnd,
+		streamReader:   streamReader,
+	}
+}
+
+type blockStreamingQuerierSeries struct {
+	labels                       labels.Labels
+	seriesIdxStart, seriesIdxEnd int
+	streamReader                 chunkStreamReader
+}
+
+func (bqs *blockStreamingQuerierSeries) Labels() labels.Labels {
+	return bqs.labels
+}
+
+func (bqs *blockStreamingQuerierSeries) Iterator(reuse chunkenc.Iterator) chunkenc.Iterator {
+	// Fetch the chunks from the stream.
+	var allChunks []storepb.AggrChunk
+	for i := bqs.seriesIdxStart; i <= bqs.seriesIdxEnd; i++ {
+		chks, err := bqs.streamReader.GetChunks(uint64(i))
+		if err != nil {
+			return series.NewErrIterator(err)
+		}
+		allChunks = append(allChunks, chks...)
+	}
+	if len(allChunks) == 0 {
+		// should not happen in practice, but we have a unit test for it
+		return series.NewErrIterator(errors.New("no chunks"))
+	}
+
+	sort.Slice(allChunks, func(i, j int) bool {
+		return allChunks[i].MinTime < allChunks[j].MinTime
+	})
+
+	it, err := newBlockQuerierSeriesIterator(reuse, bqs.Labels(), allChunks)
+	if err != nil {
+		return series.NewErrIterator(err)
+	}
+
+	return it
+}
+
+// storeGatewayStreamReader is responsible for managing the streaming of chunks from a storegateway and buffering
+// chunks in memory until they are consumed by the PromQL engine.
+type storeGatewayStreamReader struct {
+	client              storegatewaypb.StoreGateway_SeriesClient
+	expectedSeriesCount int
+	queryLimiter        *limiter.QueryLimiter
+	stats               *stats.Stats
+	log                 log.Logger
+
+	seriesChunksChan chan *storepb.StreamingChunksBatch
+	chunksBatch      []*storepb.StreamingChunks
+	errorChan        chan error
+}
+
+func newStoreGatewayStreamReader(client storegatewaypb.StoreGateway_SeriesClient, expectedSeriesCount int, queryLimiter *limiter.QueryLimiter, stats *stats.Stats, log log.Logger) *storeGatewayStreamReader {
+	return &storeGatewayStreamReader{
+		client:              client,
+		expectedSeriesCount: expectedSeriesCount,
+		queryLimiter:        queryLimiter,
+		stats:               stats,
+		log:                 log,
+		seriesChunksChan:    make(chan *storepb.StreamingChunksBatch, 1),
+		// Important: to ensure that the goroutine does not become blocked and leak, the goroutine must only ever write to errorChan at most once.
+		errorChan: make(chan error, 1),
```

Comment:
Is there a reason why we initialise these early, rather than in `StartBuffering`? If `StartBuffering` is never called, then if `GetChunks` is called, it'll block forever.

Reply:
I think I asked Ganesh to move them. I thought that initializing them in the constructor made sense (that's what constructors are for, I guess)

Calling GetChunks without StartBuffering sounds like invalid usage though.

Reply:
>Calling GetChunks without StartBuffering sounds like invalid usage though.

It would be, but I'd prefer that we don't block forever if it does happen - we'd at least get an error that would lead us to this area of the codebase.

Reply:
I can move it back